### PR TITLE
GH-35089: [CI][C++][Flight] Test failures in macos release verification nightlies

### LINKED ISSUE #35089: [CI][C++][Flight] Test failures in macos release verification nightlies
### Describe the bug, including details regarding any error messages, version, and platform.

Examples can  be found below:

https://github.com/ursacomputing/crossbow/actions/runs/4232843320/jobs/7353066727

https://github.com/ursacomputing/crossbow/actions/runs/4583238411/jobs/8093963787

### Component(s)

C++, Continuous Integration, FlightRPC, Release

### PR DESCRIPTION
**Pull request by zeroshade**
* Closes: #35089

### CODE REVIEW

**Comment by zeroshade** [comment]
@github-actions crossbow submit verify-rc-source-cpp-macos-*

**Comment by zeroshade** [comment]
@github-actions crossbow submit verify-rc-source-cpp-macos-arm64

File: cpp/src/arrow/flight/flight_test.cc [hunk 1]
```diff
@@ -1023,7 +1023,11 @@ TEST_F(TestFlightClient, TimeoutFires) {
   Status status = client->GetFlightInfo(options, FlightDescriptor{}).status();
   auto end = std::chrono::system_clock::now();
 #ifdef ARROW_WITH_TIMING_TESTS
+#ifdef __APPLE__
+  EXPECT_LE(end - start, std::chrono::milliseconds{1000});
+#else
   EXPECT_LE(end - start, std::chrono::milliseconds{400});
+#endif
```
Comment (lidavidm) [inline cpp/src/arrow/flight/flight_test.cc:1030]: Maybe just raise the timeout universally?

File: cpp/src/arrow/flight/test_definitions.cc [hunk 2]
```diff
@@ -108,7 +108,15 @@ void ConnectivityTest::TestBrokenConnection() {
   ASSERT_OK(server->Shutdown());
   ASSERT_OK(server->Wait());
 
+#if defined(__APPLE__) && defined(__x86_64__)
+  // for some reason, on macos-x86-64 the broken connection here returns
+  // a different error and we get UnknownError instead of IOError.
+  // Since we're *testing* the broken connection, we can just check for
+  // the other error in this case.
+  ASSERT_RAISES(UnknownError, client->GetFlightInfo(FlightDescriptor::Command("")));
```
Comment (lidavidm) [inline cpp/src/arrow/flight/test_definitions.cc:116]: It seems inconsistent - maybe just get the status, assert that it fails, and then `ASSERT_THAT(status.code(), ::testing::AnyOf())`?

File: cpp/src/arrow/flight/flight_test.cc (hunk 1 above)
Comment (zeroshade) [reply cpp/src/arrow/flight/flight_test.cc]: done

File: cpp/src/arrow/flight/test_definitions.cc (hunk 2 above)
Comment (zeroshade) [reply cpp/src/arrow/flight/test_definitions.cc]: done

**Comment by zeroshade** [comment]
I've reduced the failures at least, but I can't seem to figure out the cause of these Segfaults in the macos-amd64-conda release verification. Any assistance here would be amazing. Thanks!

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
However we can/should get this merged

**Comment by lidavidm** [comment]
Ok, I checked out the instance that Weston set up. (Thanks a lot Weston!)

The test fails probabilistically if I just try to run `GrpcDataTest.TestDoGetInts`. The backtrace shows this suspicious behavior: The main thread is executing a gRPC-internal destructor, `ExecCtx::~ExecCtx`:

```
  thread #1, queue = 'com.apple.main-thread'
    frame #0: 0x00007ff800bcb889 libsystem_pthread.dylib`pthread_mutex_lock + 11
    frame #1: 0x00000001002af7b9 libgpr.29.dylib`gpr_mu_lock + 9
    frame #2: 0x000000010123390f libgrpc.29.dylib`fd_orphan(grpc_fd*, grpc_closure*, int*, char const*) + 255
    frame #3: 0x0000000101249487 libgrpc.29.dylib`deactivated_all_ports(grpc_tcp_server*) + 135
    frame #4: 0x0000000101249195 libgrpc.29.dylib`on_read(void*, absl::lts_20230125::Status) + 1477
    frame #5: 0x000000010123794c libgrpc.29.dylib`grpc_core::ExecCtx::Flush() + 156
    frame #6: 0x0000000101034f11 libgrpc.29.dylib`grpc_core::ExecCtx::~ExecCtx() + 33
    frame #7: 0x00000001012e9110 libgrpc.29.dylib`grpc_server_shutdown_and_notify + 288
    frame #8: 0x0000000100a441bc libgrpc++.1.51.dylib`grpc::Server::ShutdownInternal(gpr_timespec) + 236
    frame #9: 0x00000001008d249f libarrow_flight.1200.dylib`arrow::flight::transport::grpc::(anonymous namespace)::GrpcServerTransport::Shutdown() + 47
```

Meanwhile, one of the gRPC server threads is executing the same method:

```
* thread #19, stop reason = EXC_BAD_ACCESS (code=1, address=0x9)
  * frame #0: 0x0000000101222f64 libgrpc.29.dylib`grpc_core::StatusGetChildren(absl::lts_20230125::Status) + 68
    frame #1: 0x00000001012eeb8d libgrpc.29.dylib`grpc_error_has_clear_grpc_status(absl::lts_20230125::Status) + 125
    frame #2: 0x000000010113142c libgrpc.29.dylib`close_transport_locked(grpc_chttp2_transport*, absl::lts_20230125::Status) + 92
    frame #3: 0x0000000101135173 libgrpc.29.dylib`read_action_locked(void*, absl::lts_20230125::Status) + 1491
    frame #4: 0x0000000101231afe libgrpc.29.dylib`grpc_combiner_continue_exec_ctx() + 158
    frame #5: 0x00000001012378fe libgrpc.29.dylib`grpc_core::ExecCtx::Flush() + 78
    frame #6: 0x0000000101234a06 libgrpc.29.dylib`pollset_work(grpc_pollset*, grpc_pollset_worker**, grpc_core::Timestamp) + 2662
    frame #7: 0x0000000101236f86 libgrpc.29.dylib`pollset_work(grpc_pollset*, grpc_pollset_worker**, grpc_core::Timestamp) + 22
    frame #8: 0x000000010123ad23 libgrpc.29.dylib`grpc_pollset_work(grpc_pollset*, grpc_pollset_worker**, grpc_core::Timestamp) + 19
    frame #9: 0x00000001012dedb0 libgrpc.29.dylib`cq_next(grpc_completion_queue*, gpr_timespec, void*) + 544
    frame #10: 0x0000000100a29b00 libgrpc++.1.51.dylib`grpc::CompletionQueue::AsyncNextInternal(void**, bool*, gpr_timespec) + 80
    frame #11: 0x0000000100a45655 libgrpc++.1.51.dylib`grpc::Server::SyncRequestThreadManager::PollForWork(void**, bool*) + 101
    frame #12: 0x0000000100a50700 libgrpc++.1.51.dylib`grpc::ThreadManager::MainWorkLoop() + 64
```

That said, this is supposed to be a thread local, so I don't see how they'd trample each other here. I also can't get lldb to print out the thread-local value so I can't check if it's initialized or not.

**Comment by lidavidm** [comment]
Ah, this gets lldb to recognize it:

```
(lldb) image lookup -v -s grpc_core::ExecCtx::exec_ctx_ libgrpc.29.dylib
1 symbols match 'grpc_core::ExecCtx::exec_ctx_' in /private/tmp/verify-tmp/mambaforge/envs/conda-source/lib/libgrpc.29.dylib:
        Address: libgrpc.29.dylib[0x00000000003db9b8] (libgrpc.29.dylib.__DATA.__thread_vars + 264)
        Summary: libgrpc.29.dylib`grpc_core::ExecCtx::exec_ctx_
         Module: file = "/private/tmp/verify-tmp/mambaforge/envs/conda-source/lib/libgrpc.29.dylib", arch = "x86_64"
         Symbol: id = {0x000036f2}, range = [0x00000001014019b8-0x00000001014019d0), name="grpc_core::ExecCtx::exec_ctx_", mangled="_ZN9grpc_core7ExecCtx9exec_ctx_E"

(lldb) print _ZN9grpc_core7ExecCtx9exec_ctx_E
(void *) $0 = 0x00007ff800bd743c
```

lldb says they're the same:

```
(lldb) t 1
* thread #1, queue = 'com.apple.main-thread'
    frame #0: 0x00007ff800bcb889 libsystem_pthread.dylib`pthread_mutex_lock + 11
libsystem_pthread.dylib`pthread_mutex_lock:
->  0x7ff800bcb889 <+11>: jne    0x7ff800bcb8fe            ; <+128>
    0x7ff800bcb88b <+13>: movl   0xc(%rdi), %eax
    0x7ff800bcb88e <+16>: movl   %eax, %ecx
    0x7ff800bcb890 <+18>: andl   $0x1c0, %ecx              ; imm = 0x1C0
(lldb) print _ZN9grpc_core7ExecCtx9exec_ctx_E
(void *) $1 = 0x00007ff800bd743c
```

But I wonder if it recognizes thread locals properly. 

That said, this seems to mostly be a gRPC problem...

**Comment by lidavidm** [comment]
I do see that grpc-cpp is 1.51.1 in the env and in the crashing CI build. (The brew CI build fails, but doesn't crash.) But conda-forge has 1.52 now - maybe we should try that?

**Comment by lidavidm** [comment]
Aha, and we were artificially on an old version because the conda package was renamed `grpc-cpp` to `libgrpc`.

**Comment by lidavidm** [comment]
Unfortunately, gRPC 1.52.1 has the same issue. `Illegal keepalive throttling value` also starts appearing. The stack traces look the same.

**Comment by lidavidm** [comment]
Correction: the main thread has a different stack trace, so it seems less likely that two threads are trampling on ExecCtx and more that there's something being tickled in gRPC:

```
* thread #1, queue = 'com.apple.main-thread'
  * frame #0: 0x0000000101c4c760 libgrpc.30.dylib`grpc_tcp_server_unref(grpc_tcp_server*)
    frame #1: 0x0000000101b2205e libgrpc.30.dylib`grpc_core::(anonymous namespace)::Chttp2ServerListener::Orphan() + 286
    frame #2: 0x0000000101ceb844 libgrpc.30.dylib`grpc_core::Server::ShutdownAndNotify(grpc_completion_queue*, void*) + 820
    frame #3: 0x0000000101cee7e4 libgrpc.30.dylib`grpc_server_shutdown_and_notify + 276
    frame #4: 0x0000000100daa9cc libgrpc++.1.52.dylib`grpc::Server::ShutdownInternal(gpr_timespec) + 220
    frame #5: 0x00000001012706d1 libarrow_flight.1200.dylib`arrow::flight::transport::grpc::(anonymous namespace)::GrpcServerTransport::Shutdown() + 33
    frame #6: 0x0000000101250755 libarrow_flight.1200.dylib`arrow::flight::FlightServerBase::Shutdown(std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l> > > const*) + 149
    frame #7: 0x0000000100396131 libarrow_flight_testing.1200.dylib`arrow::flight::DataTest::TearDownTest() + 129
```

**Comment by zeroshade** [comment]
@lidavidm When I followed the crossbow builds back I can't see an environmental change, the last successful commit before this started failing was https://github.com/apache/arrow/compare/f2d632e727d4eb7e8a693354c9116ea2518e50a6...1d74483fa0659aebc0cb1dfb771ba38800166bf2 on Feb 20th (which is before the last update to `grpc-cpp` in conda). 

I'm currently building grpc from source on the macos instance that @westonpace spun up to compare against the conda env. If i'm able to figure out anything I'll comment back on here.

**Comment by zeroshade** [comment]
@lidavidm I was able to reproduce these crashes by building grpc v1.51.1 from source on the macos box that @westonpace spun up while keeping the rest of the dependencies coming from Conda.

In addition, weston found this: https://github.com/conda-forge/grpc-cpp-feedstock/issues/281 which could likely explain the issue (assuming we're not missing something). I'm currently building v1.54.0 from source to confirm for myself that I didn't screw any environmental stuff up, but it looks like this explains the problem. I can try pinning conda to v1.50.0 or something instead and seeing if that alleviates the issue afterwards.

**Comment by zeroshade** [comment]
@github-actions crossbow submit verify-rc-source-cpp-macos-conda-amd64

**Comment by zeroshade** [comment]
Pinning the conda version of grpc-cpp to <=1.50.1 looks like it worked on the macosx ec2 box. So here's hoping it works for crossbow also, if so we can call this done and set! :smile:

**Comment by lidavidm** [comment]
> @lidavidm I was able to reproduce these crashes by building grpc v1.51.1 from source on the macos box that @westonpace spun up while keeping the rest of the dependencies coming from Conda.
> 
> In addition, weston found this: https://github.com/conda-forge/grpc-cpp-feedstock/issues/281 which could likely explain the issue (assuming we're not missing something). I'm currently building v1.54.0 from source to confirm for myself that I didn't screw any environmental stuff up, but it looks like this explains the problem. I can try pinning conda to v1.50.0 or something instead and seeing if that alleviates the issue afterwards.

Wow, that's an incredible find. Thanks for figuring this out!

**Comment by assignUser** [comment]
Wow, amazing job!

**Pull request merged**